File: snippets/security/authentication/multi_factor_auth_totp.ts
/**
 * Multi-Factor Authentication (TOTP) Implementation (TypeScript)
 * Production-ready 2FA with Time-based One-Time Password
 */
import * as speakeasy from 'speakeasy';
import * as QRCode from 'qrcode';
import { randomBytes } from 'crypto';

interface TOTPSecret {
  secret: string;
  qrCodeUrl: string;
  backupCodes: string[];
}

interface TOTPVerification {
  isValid: boolean;
  delta?: number;
}

export class TOTPAuthenticator {
  private issuerName: string;

  /**
   * Initialize TOTP authenticator
   * @param issuerName - Application name
   */
  constructor(issuerName: string = 'MyApp') {
    this.issuerName = issuerName;
  }

  /**
   * Generate a new TOTP secret
   * @returns Base32-encoded secret key
   */
  generateSecret(): string {
    const secret = speakeasy.generateSecret({
      length: 32,
      name: this.issuerName,
    });

    return secret.base32;
  }

  /**
   * Get provisioning URI for QR code
   * @param secret - TOTP secret key
   * @param userIdentifier - User email or username
   * @param issuerName - Optional issuer name override
   * @returns Provisioning URI
   */
  getProvisioningUri(
    secret: string,
    userIdentifier: string,
    issuerName?: string
  ): string {
    return speakeasy.otpauthURL({
      secret,
      label: userIdentifier,
      issuer: issuerName || this.issuerName,
      encoding: 'base32',
    });
  }

  /**
   * Generate QR code as data URL
   * @param secret - TOTP secret key
   * @param userIdentifier - User email or username
   * @returns Base64-encoded QR code image
   */
  async generateQRCode(
    secret: string,
    userIdentifier: string
  ): Promise<string> {
    const uri = this.getProvisioningUri(secret, userIdentifier);

    try {
      const qrCodeDataUrl = await QRCode.toDataURL(uri);
      return qrCodeDataUrl;
    } catch (error) {
      console.error('QR code generation error:', error);
      throw new Error('Failed to generate QR code');
    }
  }

  /**
   * Verify TOTP token
   * @param secret - TOTP secret key
   * @param token - 6-digit TOTP code from user
   * @param validWindow - Number of time steps to check (default 1 = 30s window)
   * @returns Verification result
   */
  verifyToken(
    secret: string,
    token: string,
    validWindow: number = 1
  ): TOTPVerification {
    const result = speakeasy.totp.verify({
      secret,
      encoding: 'base32',
      token,
      window: validWindow,
    });

    if (typeof result === 'boolean') {
      return { isValid: result };
    }

    // Some versions return delta
    return {
      isValid: true,
      delta: result as number,
    };
  }

  /**
   * Get current TOTP token (for testing)
   * @param secret - TOTP secret key
   * @returns Current 6-digit token
   */
  getCurrentToken(secret: string): string {
    return speakeasy.totp({
      secret,
      encoding: 'base32',
    });
  }

  /**
   * Get seconds remaining until next token
   * @returns Seconds until token expires
   */
  getTimeRemaining(): number {
    const now = Math.floor(Date.now() / 1000);
    return 30 - (now % 30);
  }

  /**
   * Setup 2FA for a user
   * @param userIdentifier - User email or username
   * @returns TOTP setup data
   */
  async setupTOTP(userIdentifier: string): Promise<TOTPSecret> {
    const secret = this.generateSecret();
    const qrCodeUrl = await this.generateQRCode(secret, userIdentifier);
    const backupCodes = BackupCodes.generateBackupCodes();

    return {
      secret,
      qrCodeUrl,
      backupCodes,
    };
  }
}

export class BackupCodes {
  /**
   * Generate backup recovery codes
   * @param count - Number of codes to generate
   * @returns Array of backup codes
   */
  static generateBackupCodes(count: number = 10): string[] {
    const codes: string[] = [];
    const charset = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';

    for (let i = 0; i < count; i++) {
      let code = '';
      const randomValues = randomBytes(8);

      for (let j = 0; j < 8; j++) {
        code += charset[randomValues[j] % charset.length];
      }

      // Format as XXXX-XXXX for readability
      const formatted = `${code.substring(0, 4)}-${code.substring(4)}`;
      codes.push(formatted);
    }

    return codes;
  }

  /**
   * Hash backup code for storage
   * @param code - Plain backup code
   * @returns Hashed code
   */
  static hashBackupCode(code: string): string {
    const { createHash } = require('crypto');
    return createHash('sha256').update(code).digest('hex');
  }

  /**
   * Verify backup code
   * @param code - Plain code from user
   * @param hashedCode - Stored hashed code
   * @returns True if code matches
   */
  static verifyBackupCode(code: string, hashedCode: string): boolean {
    const { createHash, timingSafeEqual } = require('crypto');
    const hash = createHash('sha256').update(code).digest();
    const storedHash = Buffer.from(hashedCode, 'hex');

    if (hash.length !== storedHash.length) {
      return false;
    }

    return timingSafeEqual(hash, storedHash);
  }
}

export class MFAManager {
  private authenticator: TOTPAuthenticator;
  private mfaData: Map<number, MFAUserData>;

  constructor(issuerName: string = 'MyApp') {
    this.authenticator = new TOTPAuthenticator(issuerName);
    this.mfaData = new Map();
  }

  /**
   * Enable MFA for a user
   * @param userId - User ID
   * @param userIdentifier - User email/username
   * @returns TOTP setup data
   */
  async enableMFA(
    userId: number,
    userIdentifier: string
  ): Promise<TOTPSecret> {
    const setup = await this.authenticator.setupTOTP(userIdentifier);

    // Store MFA data (in production, save to database)
    const backupCodesHashed = setup.backupCodes.map((code) =>
      BackupCodes.hashBackupCode(code)
    );

    this.mfaData.set(userId, {
      secret: setup.secret,
      backupCodes: backupCodesHashed,
      isEnabled: false, // Enable after verification
      createdAt: new Date().toISOString(),
    });

    return setup;
  }

  /**
   * Verify and activate MFA
   * @param userId - User ID
   * @param token - TOTP token
   * @returns True if verified and activated
   */
  verifyAndActivate(userId: number, token: string): boolean {
    const userData = this.mfaData.get(userId);

    if (!userData) {
      return false;
    }

    const verification = this.authenticator.verifyToken(
      userData.secret,
      token
    );

    if (verification.isValid) {
      userData.isEnabled = true;
      return true;
    }

    return false;
  }

  /**
   * Verify MFA code
   * @param userId - User ID
   * @param code - TOTP or backup code
   * @returns True if valid
   */
  verifyMFA(userId: number, code: string): boolean {
    const userData = this.mfaData.get(userId);

    if (!userData || !userData.isEnabled) {
      return false;
    }

    // Try TOTP first
    const totpVerification = this.authenticator.verifyToken(
      userData.secret,
      code
    );

    if (totpVerification.isValid) {
      return true;
    }

    // Try backup codes
    for (let i = 0; i < userData.backupCodes.length; i++) {
      if (BackupCodes.verifyBackupCode(code, userData.backupCodes[i])) {
        // Remove used backup code
        userData.backupCodes.splice(i, 1);
        return true;
      }
    }

    return false;
  }

  /**
   * Disable MFA for a user
   * @param userId - User ID
   * @returns True if disabled
   */
  disableMFA(userId: number): boolean {
    return this.mfaData.delete(userId);
  }

  /**
   * Check if MFA is enabled for user
   * @param userId - User ID
   * @returns True if enabled
   */
  isMFAEnabled(userId: number): boolean {
    const userData = this.mfaData.get(userId);
    return userData?.isEnabled || false;
  }
}

interface MFAUserData {
  secret: string;
  backupCodes: string[];
  isEnabled: boolean;
  createdAt: string;
}

// Example usage
if (require.main === module) {
  (async () => {
    const authenticator = new TOTPAuthenticator('MySecureApp');

    // Setup 2FA for a user
    const userEmail = 'user@example.com';
    const setup = await authenticator.setupTOTP(userEmail);

    console.log('Secret key:', setup.secret);
    console.log('QR Code:', setup.qrCodeUrl.substring(0, 50) + '...');
    console.log('Backup codes:', setup.backupCodes);

    // Get current token (for testing)
    const currentToken = authenticator.getCurrentToken(setup.secret);
    console.log('\nCurrent token:', currentToken);

    // Verify token
    const verification = authenticator.verifyToken(setup.secret, currentToken);
    console.log('Token valid:', verification.isValid);

    // Test MFA Manager
    const mfaManager = new MFAManager('MySecureApp');

    const userId = 123;
    const mfaSetup = await mfaManager.enableMFA(userId, userEmail);
    console.log('\n✓ MFA enabled for user');

    // Verify and activate
    const testToken = authenticator.getCurrentToken(mfaSetup.secret);
    const activated = mfaManager.verifyAndActivate(userId, testToken);
    console.log('MFA activated:', activated);

    // Verify MFA
    const verified = mfaManager.verifyMFA(userId, testToken);
    console.log('MFA verified:', verified);
  })();
}

File: snippets/security/authentication/multi_factor_auth_totp.py
"""
Multi-Factor Authentication (TOTP) Implementation
Production-ready 2FA with Time-based One-Time Password
"""
import pyotp
import qrcode
import io
import base64
from typing import Optional, Tuple
from datetime import datetime


class TOTPAuthenticator:
    """TOTP-based Multi-Factor Authentication"""

    def __init__(self, issuer_name: str = "MyApp"):
        self.issuer_name = issuer_name

    def generate_secret(self) -> str:
        """
        Generate a new TOTP secret key

        Returns:
            Base32-encoded secret key
        """
        return pyotp.random_base32()

    def get_provisioning_uri(
        self,
        secret: str,
        user_identifier: str,
        issuer_name: Optional[str] = None
    ) -> str:
        """
        Generate provisioning URI for QR code

        Args:
            secret: TOTP secret key
            user_identifier: User email or username
            issuer_name: Optional issuer name override

        Returns:
            Provisioning URI
        """
        totp = pyotp.TOTP(secret)
        return totp.provisioning_uri(
            name=user_identifier,
            issuer_name=issuer_name or self.issuer_name
        )

    def generate_qr_code(
        self,
        secret: str,
        user_identifier: str
    ) -> str:
        """
        Generate QR code as base64 string

        Args:
            secret: TOTP secret key
            user_identifier: User email or username

        Returns:
            Base64-encoded QR code image
        """
        uri = self.get_provisioning_uri(secret, user_identifier)

        # Generate QR code
        qr = qrcode.QRCode(version=1, box_size=10, border=5)
        qr.add_data(uri)
        qr.make(fit=True)

        img = qr.make_image(fill_color="black", back_color="white")

        # Convert to base64
        buffer = io.BytesIO()
        img.save(buffer, format='PNG')
        img_str = base64.b64encode(buffer.getvalue()).decode()

        return f"data:image/png;base64,{img_str}"

    def verify_token(
        self,
        secret: str,
        token: str,
        valid_window: int = 1
    ) -> bool:
        """
        Verify TOTP token

        Args:
            secret: TOTP secret key
            token: 6-digit TOTP code from user
            valid_window: Number of time steps to check (default 1 = 30s window)

        Returns:
            True if token is valid
        """
        totp = pyotp.TOTP(secret)
        return totp.verify(token, valid_window=valid_window)

    def get_current_token(self, secret: str) -> str:
        """
        Get current TOTP token (for testing)

        Args:
            secret: TOTP secret key

        Returns:
            Current 6-digit token
        """
        totp = pyotp.TOTP(secret)
        return totp.now()

    def get_time_remaining(self) -> int:
        """
        Get seconds remaining until next token

        Returns:
            Seconds until token expires
        """
        return 30 - (int(datetime.now().timestamp()) % 30)


class BackupCodes:
    """Backup codes for account recovery"""

    @staticmethod
    def generate_backup_codes(count: int = 10) -> list:
        """
        Generate backup recovery codes

        Args:
            count: Number of codes to generate

        Returns:
            List of backup codes
        """
        import secrets
        codes = []
        for _ in range(count):
            # Generate 8-character alphanumeric code
            code = ''.join(
                secrets.choice('ABCDEFGHJKLMNPQRSTUVWXYZ23456789')
                for _ in range(8)
            )
            # Format as XXXX-XXXX for readability
            formatted = f"{code[:4]}-{code[4:]}"
            codes.append(formatted)
        return codes

    @staticmethod
    def hash_backup_code(code: str) -> str:
        """
        Hash backup code for storage

        Args:
            code: Plain backup code

        Returns:
            Hashed code
        """
        import hashlib
        return hashlib.sha256(code.encode()).hexdigest()

    @staticmethod
    def verify_backup_code(code: str, hashed_code: str) -> bool:
        """
        Verify backup code

        Args:
            code: Plain code from user
            hashed_code: Stored hashed code

        Returns:
            True if code matches
        """
        import hashlib
        return hashlib.sha256(code.encode()).hexdigest() == hashed_code


# Example usage
if __name__ == "__main__":
    authenticator = TOTPAuthenticator(issuer_name="MySecureApp")

    # Setup 2FA for a user
    user_email = "user@example.com"
    secret = authenticator.generate_secret()
    print(f"Secret key: {secret}")

    # Generate QR code
    qr_code = authenticator.generate_qr_code(secret, user_email)
    print(f"QR Code (base64): {qr_code[:50]}...")

    # Get current token (for testing)
    current_token = authenticator.get_current_token(secret)
    print(f"Current token: {current_token}")

    # Verify token
    is_valid = authenticator.verify_token(secret, current_token)
    print(f"Token valid: {is_valid}")

    # Generate backup codes
    backup_codes = BackupCodes.generate_backup_codes()
    print(f"\nBackup codes: {backup_codes}")

    # Hash backup codes for storage
    hashed_codes = [BackupCodes.hash_backup_code(code) for code in backup_codes]
    print(f"First hashed code: {hashed_codes[0]}")
